ext/node/ops/tls.rs

```diff
@@ -68,3 +102,492 @@ pub fn op_tls_canonicalize_ipv4_address(
+  hostname: String,
+  alpn_protocols: Option<Vec<String>>,
+  reject_unauthorized: Option<bool>,
+}
+
+#[derive(Debug)]
+pub struct JSStreamTlsResource {
+  rd: AsyncRefCell<TlsStreamRead<JSStreamSocket>>,
+  wr: AsyncRefCell<TlsStreamWrite<JSStreamSocket>>,
+}
+
+impl JSStreamTlsResource {
+  pub fn new(
+    (rd, wr): (
+      TlsStreamRead<JSStreamSocket>,
+      TlsStreamWrite<JSStreamSocket>,
+    ),
+  ) -> Self {
+    Self {
+      rd: AsyncRefCell::new(rd),
+      wr: AsyncRefCell::new(wr),
+    }
+  }
+
+  pub async fn handshake(
+    self: &Rc<Self>,
+  ) -> Result<TlsHandshakeInfo, std::io::Error> {
+    let mut wr = RcRef::map(self, |r| &r.wr).borrow_mut().await;
+
+    let handshake = wr.handshake().await?;
+
+    let alpn_protocol = handshake.alpn.map(|alpn| alpn.into());
+    let peer_certificates = handshake.peer_certificates.clone();
+    let tls_info = TlsHandshakeInfo {
+      alpn_protocol,
+      peer_certificates,
+    };
+
+    Ok(tls_info)
+  }
+
+  pub async fn read(
+    self: Rc<Self>,
+    data: &mut [u8],
+  ) -> Result<usize, std::io::Error> {
+    use tokio::io::AsyncReadExt;
+
+    let mut rd = RcRef::map(&self, |r| &r.rd).borrow_mut().await;
+    rd.read(data).await
+  }
+
+  pub async fn write(
+    self: Rc<Self>,
+    data: &[u8],
+  ) -> Result<usize, std::io::Error> {
+    use tokio::io::AsyncWriteExt;
+
+    let mut wr = RcRef::map(&self, |r| &r.wr).borrow_mut().await;
+    let nwritten = wr.write(data).await?;
+    wr.flush().await?;
+    Ok(nwritten)
+  }
+}
+
+impl Resource for JSStreamTlsResource {
+  deno_core::impl_readable_byob!();
+  deno_core::impl_writable!();
+
+  fn name(&self) -> Cow<'_, str> {
+    "JSStreamTlsResource".into()
+  }
+}
+
+#[op2]
+pub fn op_node_tls_start(
+  state: Rc<RefCell<OpState>>,
+  #[serde] args: StartJSTlsArgs,
+  #[buffer] output: &mut [u32],
+) -> Result<(), NetError> {
+  let reject_unauthorized = args.reject_unauthorized.unwrap_or(true);
+  let hostname = match &*args.hostname {
+    "" => "localhost".to_string(),
+    n => n.to_string(),
+  };
+
+  assert_eq!(output.len(), 2);
+
+  let ca_certs = args
+    .ca_certs
+    .into_iter()
+    .map(|s| s.into_bytes())
+    .collect::<Vec<_>>();
+
+  let hostname_dns = ServerName::try_from(hostname.to_string())
+    .map_err(|_| NetError::InvalidHostname(hostname))?;
+  // --unsafely-ignore-certificate-errors overrides the `rejectUnauthorized` option.
+  let unsafely_ignore_certificate_errors = if reject_unauthorized {
+    state
+      .borrow()
+      .try_borrow::<UnsafelyIgnoreCertificateErrors>()
+      .and_then(|it| it.0.clone())
+  } else {
+    Some(Vec::new())
+  };
+
+  let root_cert_store = state
+    .borrow()
+    .borrow::<DefaultTlsOptions>()
+    .root_cert_store()
+    .map_err(NetError::RootCertStore)?;
+
+  let (network_to_tls_tx, network_to_tls_rx) =
+    tokio::sync::mpsc::channel::<Bytes>(10);
+  let (tls_to_network_tx, tls_to_network_rx) =
+    tokio::sync::mpsc::channel::<Bytes>(10);
```

Comment:
Are these totally random buffer sizes?

Reply:
Yeah, it would never have more than 1 message in queue though.